lib/tests/streamlit/runtime/app_session_test.py

```diff
@@ -80,14 +80,28 @@ def _create_test_session(
         "streamlit.runtime.app_session.LocalSourcesWatcher",
         MagicMock(spec=LocalSourcesWatcher),
     ):
-        return AppSession(
+        session = AppSession(
             script_data=ScriptData("/fake/script_path.py", is_hello=False),
             uploaded_file_manager=MagicMock(),
             script_cache=MagicMock(),
             message_enqueued_callback=None,
             user_info={"email": "test@example.com"},
             session_id_override=session_id_override,
         )
+        import sys
+
+        # in Python 3.13, the "test_shutdown" test is failing because shutdown is called
+        # twice and not None. I was able to figure that this is because the
+        # AppSession.__del__ method is called which calls shutdown() and disconnect
+        # again. I am not sure what changed in 3.13 - perhaps a session object of
+        # another test is garbage collected quicker which calls the __del__ method.
+        # Another alternative is to move the test_shutdown test to its own class
+        # to avoid this side-effect. There is no test to rely on the __del__ method
+        # call, so for now this side-effect is disabled completely.
```

Comment:
**question:** Just to be sure I'm understanding this comment block correctly, this is stating that the issue is purely an issue that is coming up in tests, correct? Per the last suggestion, if we moved these tests into their own class would that also allow us to remove the conditional logic for the version of Python?

Reply:
Yep, this is only a test-issue because the test counts the number of times disconnect is called. However, in reality calling it more than once is a no-op because it is guarded by a flag [here](https://github.com/streamlit/streamlit/blob/8cb5af07743196950aae88a54db7ad327c56a1b0/lib/streamlit/runtime/app_session.py#L240).
We would get rid of the version check yeah! Do you think that would be nicer?

Reply:
Thanks for the clarity. My perspective is that tests should accurately reflect the expected behavior. Since it sounds like this as written works around a test issue, I'd personally opt for a solution that gives us higher confidence. That could be a different class, or even reconsidering how we're testing this code here since this assertion may not be congruent with the expected runtime behavior given there is a guard. What do you think is the best course of action given these thoughts?

Reply:
@sfc-gh-bnisco Your thoughts have prompted me to spend more time and look more deeply into this and I have fixed the underlying issue! Apparently, there was some issue with the mock of `uploaded_file_mgr=MagicMock()` which I have changed to `uploaded_file_manager=MagicMock(spec=UploadedFileManager),`. Now, the tests seem to run through perfectly fine. It's curious that this issue only surfaced in Python 3.13. The docs of `MagicMock` state that `If you use the spec or spec_set arguments then only magic methods that exist in the spec will be created.` and it looks like the mock was somewhat wrong previously; although I am still not 100% sure what in the mocking caused this behavior / race condition(?).